File: src/my_py_pkg/my_py_pkg/add_two_ints_client.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from functools import partial
# functools.partial: 
# Cho phép tạo một phiên bản mới của một hàm với một số đối số mặc định đã được cung cấp.

from example_interfaces.srv import AddTwoInts

class AddTwoIntsClientNode(Node):
    def __init__(self):
        super().__init__("add_two_ints_client") # modify name
        self.call_add_two_ints_server(6,3)
        self.call_add_two_ints_server(5,5)
        self.call_add_two_ints_server(10,10)

    def call_add_two_ints_server(self, a, b):
        client = self.create_client(AddTwoInts, "add_two_ints")
        while not client.wait_for_service(1.0):
            self.get_logger().warn("Waiting for Server Add Two Ints")
        request = AddTwoInts.Request()
        request.a = a
        request.b = b
        future = client.call_async(request=request)
        # Add a callback to be executed when the task is done
        future.add_done_callback(partial(self.callback_call_add_two_ints,a=a, b=b))
    
    def callback_call_add_two_ints(self, future, a, b):
        try:
            response = future.result()
            self.get_logger().info(str(a) + " + " + 
                                str(b) + " = " + str(response.sum))
        except Exception as e:
            self.get_logger().error("Service call failed !")


def main(args = None):
    rclpy.init(args=args)
    node = AddTwoIntsClientNode() # modify name
    rclpy.spin(node=node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: src/my_cpp_pkg/src/add_two_ints_client.cpp
#include "rclcpp/rclcpp.hpp"
# include "example_interfaces/srv/add_two_ints.hpp"

class AddTwoIntsClientNode: public rclcpp::Node  // mofify name
{
    public:
        AddTwoIntsClientNode(): Node("cpp_test") // modify name
        {
            thread1_ = std::thread(std::bind(&AddTwoIntsClientNode::callAddTwoIntsService, this, 1, 4));
            // send multiple to the server
            threads_.push_back(std::thread(std::bind(&AddTwoIntsClientNode::callAddTwoIntsService, this, 1, 4)));
            

        }

        void callAddTwoIntsService(int a, int b)
        {
            auto client = this->create_client<example_interfaces::srv::AddTwoInts>("add_two_ints");
            while (!client->wait_for_service(std::chrono::seconds(1)))
            {
                RCLCPP_WARN(this->get_logger(), "Waiting for the server to be up ...");
            }

            auto request = std::make_shared<example_interfaces::srv::AddTwoInts::Request>();
            request->a = a;
            request->b = b;

            auto future = client->async_send_request(request);

            try{
                auto respone = future.get();
                RCLCPP_INFO(this->get_logger(), "%ld + %ld = %ld", request->a, request->b, respone->sum);

            }catch(const std::exception &e)
            {
                RCLCPP_ERROR(this->get_logger(), "Service call failed !");
            }
        }

    private:
        std::thread thread1_;
        // multiple send
        std::vector<std::thread> threads_;
};
int main(int argc, char **argv)
{
    rclcpp::init(argc,argv);
    auto node = std::make_shared<AddTwoIntsClientNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}